src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,372 @@
+        TextFormat::ParseFromString(kSuccessResponse1, &pass_quota_response1_));
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kErrorResponse1, &error_quota_response1_));
+  }
+
+  AllocateQuotaRequest quota_request1_;
+  AllocateQuotaResponse pass_quota_response1_;
+  AllocateQuotaResponse error_quota_response1_;
+
+  MockCheckTransport mock_check_transport_;
+  MockQuotaTransport mock_quota_transport_;
+  MockReportTransport mock_report_transport_;
+};
+
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestNonCachedQuotaWithStoredCallback) {
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(0, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  int callbackCalledCount = 0;
+  options.quota_transport = [this, &callbackCalledCount](
+      const AllocateQuotaRequest& request, AllocateQuotaResponse* response,
+      TransportDoneFunc on_done) {
+    response = &pass_quota_response1_;
+
+    callbackCalledCount++;
+    on_done(Status::OK);
+  };
+
+  std::unique_ptr<ServiceControlClient> client_ =
+      CreateServiceControlClient(kServiceName, kServiceConfigId, options);
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&done_status](Status status) { done_status = status; });
+  ASSERT_TRUE(TextFormat::ParseFromString(kSuccessResponse1, &quota_response));
+
+  for (int i = 0; i < 10; i++) {
+    client_->Quota(quota_request1_, &quota_response,
+                   [&done_status](Status status) { done_status = status; });
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kSuccessResponse1, &quota_response));
+  }
+  EXPECT_EQ(done_status, Status::OK);
+
+  EXPECT_EQ(callbackCalledCount, 11);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 11);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 11);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+TEST_F(ServiceControlClientImplQuotaTest, TestCachedQuotaWithStoredCallback) {
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(10, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  int callbackCalledCount = 0;
+  options.quota_transport = [&callbackCalledCount](
+      const AllocateQuotaRequest& request, AllocateQuotaResponse* response,
+      TransportDoneFunc on_done) {
+    callbackCalledCount++;
+    on_done(Status::OK);
```

Comment:
this is "InPlace",  Stored callback means you store the on_done callback,  and call it manually at a later time.